packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -2615,6 +2621,48 @@ function commitPassiveMountOnFiber(
       }
       break;
     }
+    case HostRoot: {
+      if (enableCache) {
+        // todo compare caches here
+        // if caches not same
+        // - retain next cache
+        // - if prev cache non-null: release (or move to unmount phase?)
+        // add comment that retain/release on child is technically not required
+        const previousCache: ?Cache =
+          finishedWork.alternate?.memoizedState.cache;
+        const nextCache: Cache = finishedWork.memoizedState.cache;
+        const nextCacheRetained = (nextCache: any).retained;
+        if (nextCache !== previousCache) {
+          retainCache(nextCache);
+          if (previousCache != null) {
+            releaseCache(previousCache);
+          }
+        } else if (!nextCacheRetained) {
+          retainCache(nextCache);
+          (nextCache: any).retained = true;
+        }
```

Comment:
This is a definite hack: on initial mound, `previousCache` and `nextCache` are the same value rather than `previousCache` being null, which means we can't tell that we need to retain the cache at the root. I'm setting a property on the cache instance to get around this, but this is clearly not ideal. Suggestions? @acdlite mentioned maybe threading in an extra argument to this function, but i'm not sure what I should pass in to compare against.

Reply:
Ah good point. That works for now since we don't ever cleanup the root.